ForgeUnity/Assets/BeardedManStudios/Scripts/Networking/Forge/Networking/Objects/NetworkObject.cs

```diff
@@ -832,6 +842,199 @@ public bool RegisterOnce(uint id)
+		/// <returns></returns>
+		public Task<T> SendRequest<T>(NetworkingPlayer targetPlayer, string name, Receivers receivers, object data = null)
+		{
+			if (!requestHandlerMap.ContainsKey(name))
+				throw new BaseNetworkException($"Sending request failed: request name '{name}' was not registered");
+
+			if (receivers != Receivers.Target && receivers != Receivers.Server && receivers != Receivers.Owner)
+			{
+				throw new BaseNetworkException(
+					$"Requests can only be sent to one receiver currently. If you want to send to multiple " +
+					"receivers, either send multiple requests or use the SendRequest overload using an array of " +
+					"networking players, which handles this for you."
+				);
+			}
+
+			var requestId = Interlocked.Increment(ref nextRequestId);
+			object[] args = new object[] { name, requestId, ObjectMapper.BMSByte(data).byteArr };
+			TaskCompletionSource<object> promise = new TaskCompletionSource<object>();
+
+			pendingRequests.TryAdd(requestId, promise);
+
+			BMSLog.Log($"Sending request '{name}' with ID {requestId} and data '{data}' to receivers '{receivers}'");
+
+			SendRpc(targetPlayer, "HandleRequest", false, receivers, args);
+
+			return promise.Task.ContinueWith((state) =>
+			{
+				try
+				{
+					return (T)state.Result;
+				}
+				catch (InvalidCastException e)
+				{
+					throw new BaseNetworkException(
+						$"You indicated when sending request '{name}' that the response would be of type " +
+						$"'{typeof(T)}', but the return type was set to a different type when registering the request",
+						e
+					);
+				}
+			});
+		}
+
+		/// <summary>
+		/// Registers a new request that can be sent to this object.
+		///
+		/// Requests are similar to RPCs. They only differ in that they can send back an arbitrary response and use a
+		/// TAP-based (Task-based Asynchronous Programming) async flow and are thus more convenient to use in scenarios
+		/// where you want to follow a HTTP-like or request-like flow. However, they tend to be slower because they
+		/// format incoming and outgoing data through the .NET binary formatter.
+		///
+		/// A generic object comes in. Cast it to the appropriate type you expect.
+		/// </summary>
+		/// <param name="name">The name of the request</param>
+		/// <param name="handler">The handler to execute when this request is received</param>
+		public void RegisterRequest<T1,T2>(string name, Func<RequestContext<T1>,Task<T2>> handler)
+		{
+			if (requestHandlerMap.ContainsKey(name))
+				throw new BaseNetworkException($"The request '{name}' has already been registered");
+
+			async void requestHandler(RpcArgs rpcArgs)
+			{
+				BMSLog.Log($"Received request '{name}' from {rpcArgs.Info.SendingPlayer.Ip}");
+
+				var requestId = rpcArgs.GetNext<long>();
+				var requestData = rpcArgs.GetNext<byte[]>();
+
+				BMSByte buffer = new BMSByte();
+				buffer.Clone(requestData);
```

Comment:
This may not be optimal as it copies data unnecessarily: I haven't checked if there is a way to create a `BMSByte` from an existing `byte[]` without copying.